ERCS/erc-7772.md

```diff
@@ -0,0 +1,271 @@
+        refillEntryPointDeposit(_cachedPrice);
+    }
+}
+```
+![alt text](image.png)
+
+## Rationale
+This design allows the dApp to have full control over the maximum amount of gas it wishes to sponsor. By leveraging the transient storage opcode (`tstore`), the dApp can determine and set the amount of gas it will sponsor for a given user operation.
+
+### Paymaster behavior
+During `validatePaymasterUserOp`, the Paymaster will iterate through the list of sponsors and call the authoriseGas function implemented by the dApp. The main purpose of this is for dApps that are not involved in the execution phase to also be able to set in the transient storage a specified amount they wish to sponsor.
+
+During the `postOp` phase, the Paymaster can iterate through the list of sponsors and call the `sponsorGas` function implemented by the dApp. The dApp then retrieves the sponsored amount from transient storage and transfers it to the Paymaster.
+
+In the above TokenPaymaster example, when the TokenPaymaster refunds the excess token after calculating `actualChargeNative`, it may deduct the `gasCollected` amount that was collected from the sponsors. Therefore, it will refund a greater amount to the user, essentially allowing the dApp to sponsor the transaction.
+
+### DApp behavior
+The dApp has the flexibility to decide when and how much to sponsor for the transaction. This can be done by implementing more complex logic in the code to store the amount using `tstore`.
+
+When the `authoriseGas` function is invoked, the dApp can choose to store the amount they wish to sponsor using `tstore`. If the dApp wishes to only determine the amount to sponsor during execution phase, the dApp can choose to omit the logic in this function.
+
+When the `sponsorGas` function is invoked, the dApp can transfer either the maximum amount it wishes to sponsor or the amount that the Paymaster requested, whichever is lower. 
+
+### Transient Storage
+Since all of this occurs within a single atomic transaction, the transient storage memory is not cleared until the transaction is completed, eliminating the need to re-validate `msg.sender` during `postOp`. If a specified amount is stored in transient storage but not transferred out, the storage will be cleared at the end of the transaction, ensuring that the dApp will not sponsor that particular amount.
+
+### Fallback
+#### Insufficient dApp funds
+If the dApp has insufficient funds, the try catch block will fail, and therefore the dApp will not be able to sponsor a portion of the transaction. However, this does not affect transaction execution as the paymaster would have committed to pay for the transaction. 
+
+## Backwards Compatibility
+### Existing paymaster upgrades
+Existing Paymaster contracts can be modified to support this form of gas collection from dApps. Dapps should also implement allowance setting and sponsorGas function within their existing contract.
+
+
+
+## Reference Implementation
+TBD
```

Comment:
This section is optional. You can omit it for now, and re-add it when you have something.

```suggestion
```